keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/DockerArtifactPublisher.kt

```diff
@@ -0,0 +1,104 @@
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DockerVersioningStrategy
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.api.artifacts.SpinnakerArtifact
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.BRANCH_JOB_COMMIT_BY_JOB
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.SEMVER_JOB_COMMIT_BY_JOB
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.SEMVER_JOB_COMMIT_BY_SEMVER
+import com.netflix.spinnaker.keel.api.artifacts.VersioningStrategy
+import com.netflix.spinnaker.keel.api.plugins.ArtifactPublisher
+import com.netflix.spinnaker.keel.api.plugins.SupportedArtifact
+import com.netflix.spinnaker.keel.api.plugins.SupportedVersioningStrategy
+import com.netflix.spinnaker.keel.api.support.EventPublisher
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.core.api.DEFAULT_SERVICE_ACCOUNT
+import com.netflix.spinnaker.keel.core.comparator
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import org.springframework.stereotype.Component
+
+/**
+ * Built-in keel implementation of [ArtifactPublisher] that does not itself receive/retrieve artifact information
+ * but is used by keel's `POST /artifacts/events` API to notify the core of new Docker artifacts.
+ */
+@Component
+class DockerArtifactPublisher(
+  override val eventPublisher: EventPublisher,
+  private val repository: KeelRepository,
+  private val cloudDriverService: CloudDriverService
+) : ArtifactPublisher<DockerArtifact> {
+  override val supportedArtifact = SupportedArtifact("docker", DockerArtifact::class.java)
+
+  override val supportedVersioningStrategies: List<SupportedVersioningStrategy<*>>
+    get() = listOf(
+      SupportedVersioningStrategy("docker", DockerVersioningStrategy::class.java)
+    )
+
+  override suspend fun getLatestArtifact(artifact: DeliveryArtifact): SpinnakerArtifact? {
+    if (artifact !is DockerArtifact) {
+      throw IllegalArgumentException("Only Docker artifacts are supported by this implementation.")
+    }
+
+    val serviceAccount = artifact.deliveryConfigName?.let { repository.getDeliveryConfig(it) }
+      ?.serviceAccount
+      ?: DEFAULT_SERVICE_ACCOUNT
+
+    val latestTag = cloudDriverService
+      .findDockerTagsForImage("*", artifact.name, serviceAccount)
+      .distinct()
+      .sortedWith(artifact.versioningStrategy.comparator)
+      .firstOrNull()
+
+    return if (latestTag != null) {
+      // TODO: do we need to look in a specific account? if so, then DockerArtifact should have an
+      // `account` property. In retrospect, it feels weird that we specify images without a repository
+      // URL (from which we could infer the account), or an account.
+      cloudDriverService.findDockerImages(account = "*", repository = artifact.name, tag = latestTag)
```

Comment:
@emjburns Would love your thoughts on this. It feels to me like we should add an `account` property in `DockerArtifact`. I understand we resolve the account automatically when finding Docker images to deploy, but the way we do it is to look at the Titus cluster spec, which means we can't really make sense of a Docker artifact without a cluster. I think that should not be the case.

Reply:
At netflix we have images defaulted to only one place: 
- amis are _only_ in the `test` account, then shared with other accounts
- docker images are in both the test and prod docker registries
So for our artifacts, we don't need an account specified. That's why we didn't put it in earlier.